File: src/detect-version.js
// src/detect-version.js
import { BinaryReader } from './binary-reader.js';

export class UnsupportedVersionError extends Error {
  constructor(raw) {
    super(`Unsupported Spine skeleton version: ${raw}`);
    this.name = 'UnsupportedVersionError';
    this.version = raw;
  }
}

/** Reads just the header and classifies the version. Two supported
 *  versions have genuinely different header shapes (see this task's
 *  supplement for why), so this tries the 4.2 shape first, then falls
 *  back to the 3.8 shape on a fresh reader if that doesn't parse as a
 *  4.2.x version string. Does not consume anything from the caller's own
 *  reader — always operates on its own fresh, throwaway BinaryReader(s);
 *  the caller (parseSkeleton) re-derives the right header-skip afterward
 *  once it knows which version this returned. */
export function detectVersion(bytes) {
  // Try 4.2's header shape: 8-byte (2×int32) hash, then version string.
  // A buffer too small for this shape (e.g. a short 3.8 file, whose first
  // bytes misread as a 4.2 hash encode a string length past the end of the
  // buffer) throws RangeError here — that's the signal to fall through to
  // the 3.8 shape below, not a real error.
  let raw42;
  try {
    const r42 = new BinaryReader(bytes);
    r42.readInt32(); r42.readInt32();
    raw42 = r42.readString();
  } catch (e) {
    if (!(e instanceof RangeError)) throw e;
  }
  const m42 = /^(\d+)\.(\d+)\.(\d+)/.exec(raw42 ?? '');
  if (m42 && Number(m42[1]) === 4 && Number(m42[2]) === 2) {
    return { major: 4, minor: 2, patch: Number(m42[3]), raw: raw42 };
  }

  // Fall back to 3.8's header shape: hash as a string, then version string.
  const r38 = new BinaryReader(bytes);
  r38.readString(); // hash
  const raw38 = r38.readString();
  const m38 = /^(\d+)\.(\d+)\.(\d+)/.exec(raw38 ?? '');
  if (m38 && Number(m38[1]) === 3 && Number(m38[2]) === 8) {
    return { major: 3, minor: 8, patch: Number(m38[3]), raw: raw38 };
  }

  throw new UnsupportedVersionError(raw42 ?? raw38 ?? null);
}

File: skel_mesh_parser/detect_version.py
"""Reads just the header and classifies the skeleton binary version -- port
of www/js/vendor/spine-skeleton-binary/detect-version.js's detectVersion.

Tries the 4.2 shape first (8-byte/2xint32 hash, then version string), then
falls back to the 3.8 shape (string hash, then version string) on a fresh
reader. A buffer too small for a given shape raises BufferUnderrunError --
Python's equivalent of DataView's RangeError -- which is the signal to fall
through, not a real error (same as the real JS's RangeError guard)."""
from __future__ import annotations

import re
from typing import Any

from skel_mesh_parser.binary_reader import BinaryReader, BufferUnderrunError

_VERSION_RE = re.compile(r"^(\d+)\.(\d+)\.(\d+)")


class UnsupportedVersionError(Exception):
    def __init__(self, raw: str | None) -> None:
        super().__init__(f"Unsupported Spine skeleton version: {raw}")
        self.version = raw


def detect_version(data: bytes) -> dict[str, Any]:
    raw42: str | None = None
    try:
        r42 = BinaryReader(data)
        r42.read_int32()
        r42.read_int32()
        raw42 = r42.read_string()
    except BufferUnderrunError:
        pass

    m42 = _VERSION_RE.match(raw42 or "")
    if m42 and int(m42.group(1)) == 4 and int(m42.group(2)) == 2:
        return {"major": 4, "minor": 2, "patch": int(m42.group(3)), "raw": raw42}

    raw38: str | None = None
    try:
        r38 = BinaryReader(data)
        r38.read_string()  # hash
        raw38 = r38.read_string()
    except BufferUnderrunError:
        pass

    m38 = _VERSION_RE.match(raw38 or "")
    if m38 and int(m38.group(1)) == 3 and int(m38.group(2)) == 8:
        return {"major": 3, "minor": 8, "patch": int(m38.group(3)), "raw": raw38}

    raise UnsupportedVersionError(raw42 or raw38 or None)
